### PR TITLE
Use send request

### PR DESCRIPTION
**Pull request by Maksandre**

### CODE REVIEW

File: contracts/Market.sol [hunk 1]
```diff
@@ -8,56 +8,35 @@ import {XToken} from "./XToken.sol";
...
         // TODO: if wallet does not support approve/transferFrom it should be able to transfer token directly
         require(_checkAllowanceToMarket(msg.sender, _nftId, 1), "NFT is not approved");
         s_orders[_nftId] = Order(msg.sender, address(0), _currencyId, _price, OrderState.PLACED);
+        emit Put(msg.sender, _price);
     }
 
-    function buy(uint256 _nftId) external payable {
+    function buy(CurrencyId _nftId) external payable {
         Order storage order = s_orders[_nftId];
         require(order.price > 0, "Order not found");
```
Comment (PavelMal) [inline contracts/Market.sol:31]: why doest statement 'if price > 0 => "Order not found"' mean? 
Maybe it's better to check that currencyId is empty?

File: contracts/Market.sol [hunk 2]
```diff
@@ -68,22 +47,65 @@ contract Market is NilBase {
...
 
-    function getBalance(address _owner, uint256 _token) external view returns (uint256) {
+    function handleDepositNFT(bool _success, bytes memory, bytes memory _context) public onlyResponse {
+        if (!_success) {
+            // TODO clear status
+        } else {
+            (CurrencyId nft, address buyer) = abi.decode(_context, (CurrencyId, address));
+
+            Order memory order = s_orders[nft];
+            require(s_virtualBalance[buyer][order.currencyId] >= order.price, "Buyer's virtual balance low");
```
Comment (PavelMal) [inline contracts/Market.sol:79]: if s_virtualBalance[buyer][order.currencyId] == order.price - it means that virtual balance if enough for buying order, let's change to require(s_virtualBalance[buyer][order.currencyId] > order.price,

File: contracts/test/WalletRevert.sol [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        uint256 length = _tokens.length;
+        for (uint256 i = 0; i < length; i++) {
+            Nil.Token memory tkn = _tokens[i];
+            s_allowances[_spender][tkn.id] = tkn.amount;
+            emit Approval(_spender, tkn.id, tkn.amount);
+        }
+    }
+
+    // rename to approvedTransfer? transferFrom?
+    function transfer(Nil.Token[] memory _tokens, address _recepient) public onlyInternal {
```
Comment (PavelMal) [inline contracts/test/WalletRevert.sol:37]: Do we need to Emit an event in the end of the function?

File: test/wallet.spec.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  PublicClient,
+  toHex,
+} from "@nilfoundation/niljs";
+import { expect } from "chai";
+import { XWallet, XContract } from "@spacebit/simple-nil";
+import config from "../config";
+import { Market$Type } from "../artifacts/contracts/Market.sol/Market";
+
+describe("XWallet", () => {
+  it("Can approve tokens for transfer", async () => {});
```
Comment (PavelMal) [inline test/wallet.spec.ts:15]: TODO

File: contracts/Market.sol (hunk 1 above)
Comment (Maksandre) [reply contracts/Market.sol:31]: Cannot use 0 currency because it represents a native token.

File: contracts/test/WalletRevert.sol (hunk 3 above)
Comment (Maksandre) [reply contracts/test/WalletRevert.sol:37]: Absolutely, will add this in the following pull request with all TODO fixes

File: test/wallet.spec.ts (hunk 4 above)
Comment (Maksandre) [reply test/wallet.spec.ts]: done

**Pull request merged**